termlang/termlangprolog.ml

```diff
@@ -2461,16 +2461,20 @@ and handleConstraint_mutable c =
   match c with
     `Unif(bound,p1,p2) -> pattUnifyFull_mutable ~bound:bound p1 p2
   | `UnifCanon(bound,p1,p2) -> pattcanonUnifyFull_mutable ~bound:bound p1 p2
-  | (`Demand(idx,_,p,env) as demand) ->
-    let isconcrete =
+  | (`Demand(idx,triggeridx,p,env) as demand) ->
+    let newMetaIdx =
      (match getMetaParent_mutable idx with
          Some (p, _) ->
            (match (chasePattneut_mutable [] p).term with
-               `Meta(_) -> false
-             | _ -> true)
-       | _ -> true)
+               `Meta(_, idx', _, _) -> Some(idx')
+             | _ -> None)
+       | _ -> None)
     in
-    if isconcrete then addOpenGoal_mutable p env else addConstraint_mutable idx demand
+    let isconcrete = Option.is_none newMetaIdx in
+    if isconcrete then addOpenGoal_mutable p env else (
+      let idx' = Option.get newMetaIdx in
+      addConstraint_mutable idx' (`Demand(idx',triggeridx,p,env))
+    )
```

Comment:
_Just a small nit:_
I would drop the `isconcrete`, it seems like baggage from the old code. To something like:

```ocaml
match newMetaIdx with
   Some (idx') -> addConstraint_mutable idx' (`Demand(idx',triggeridx,p,env))
 | None -> addOpenGoal_mutable p env
```

But maybe is still worth it to have something that says "if it's concrete do this, else do that". So feel free to drop the suggestion.